src/Microsoft.ML.Data/Model/ModelOperationsCatalog.cs

```diff
@@ -28,64 +28,69 @@ internal ModelOperationsCatalog(IHostEnvironment env)
             _env = env;
         }
 
+        private static TransformerChain<ITransformer> CreateChain(ITransformer transformer)
+        {
+            Contracts.AssertValueOrNull(transformer);
+            if (transformer == null)
+                return new TransformerChain<ITransformer>();
+            return new TransformerChain<ITransformer>(transformer);
+        }
+
         /// <summary>
-        /// Save the model to the stream.
+        /// Save a transformer model and the loader used to create its input data to the stream.
         /// </summary>
-        /// <param name="model">The trained model to be saved.</param>
+        /// <param name="model">The trained model to be saved. Note that this can be <see langword="null"/>, as a shorthand
+        /// for an empty data set. Upon loading with <see cref="LoadWithDataLoader(Stream, out IDataLoader{IMultiStreamSource})"/>
+        /// the returned valued will be an empty <see cref="TransformerChain{TLastTransformer}"/>.</param>
+        /// <param name="loader">The loader that was used to create data to train the model.</param>
         /// <param name="stream">A writeable, seekable stream to save to.</param>
-        public void Save<TSource>(IDataLoader<TSource> model, Stream stream)
+        public void Save<TSource>(ITransformer model, IDataLoader<TSource> loader, Stream stream)
```

Comment:
Major things on the top of my mind, the save/load methods are still inconsistent w.r.t. where they put the file or stream argument... here we see, it is the *last* argument, whereas during loading it is the *first* argument.

One possibility is we leave it as is, since in this case it is where the model is going (so maybe last is most clear), whereas during loading it is where the model is coming from.

Reply:
I kind of like this possibility.
If we're going with the "intuitive" ordering, then maybe `loader` should come before `model`?

---
In reply to: [267790407](https://github.com/dotnet/machinelearning/pull/3044#discussion_r267790407) [](ancestors = 267790407)

Reply:
Yes I considered it. Actually intuitive cuts two ways here, since this `Save` method interacts with intellisense in a way that I feel makes this arrangement most natural. But I don't insist on it.